beacon-chain/sync/data_column_sidecars.go

```diff
@@ -39,249 +39,394 @@ type DataColumnSidecarsParams struct {
+		slotsWithCommitments[block.Slot()] = true
+
+		root := roBlock.Root()
+		roots[root] = true
+		storedIndicesByRoot[root] = storage.Summary(root).Stored()
+	}
 
-		storedIndices := indicesByRootStored[root]
-		if len(storedIndices) == 0 {
+	return slotsWithCommitments, storedIndicesByRoot, roots, nil
+}
+
+// requestSidecarsFromStorage attempts to retrieve data column sidecars for each block root in `roots`
+// and for all indices specified in `requestedIndices`. If some requested sidecars are missing from storage
+// but can be reconstructed from other available sidecars, the function will try to reconstruct them.
+//
+// If not all requested sidecars can be obtained for a given root, that root is excluded from the result.
+// It returns a map from each root to its successfully retrieved sidecars.
+//
+// WARNING: This function mutates `roots` by removing entries for which all requested sidecars
+// were successfully retrieved.
+func requestSidecarsFromStorage(
+	storage filesystem.DataColumnStorageReader,
+	roots map[[fieldparams.RootLength]byte]bool,
+	storedIndicesByRoot map[[fieldparams.RootLength]byte]map[uint64]bool,
+	requestedIndicesMap map[uint64]bool,
+) (map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, error) {
+	requestedIndices := sortedSliceFromMap(requestedIndicesMap)
+
+	result := make(map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, len(roots))
+	for root := range roots {
+		storedIndices := storedIndicesByRoot[root]
+		storedIndiceCount := uint64(len(storedIndices))
+
+		// Check if reconstruction is needed.
+		reconstructionNeeded := false
+		for index := range requestedIndicesMap {
+			if !storedIndices[index] {
+				reconstructionNeeded = true
+				break
+			}
+		}
+
+		// If reconstruction is not needed, retrieve sidecars directly from storage and return them.
+		if !reconstructionNeeded {
+			verifiedRoSidecars, err := storage.Get(root, requestedIndices)
+			if err != nil {
+				return nil, errors.Wrapf(err, "storage get for block root %#x", root)
+			}
+
+			result[root] = verifiedRoSidecars
+			delete(roots, root)
+			continue
+		}
+
+		// Reconstruction is needed. Check if reconstruction is possible. If not, skip this root.
+		if storedIndiceCount < peerdas.MinimumColumnCountToReconstruct() {
 			continue
 		}
 
-		storedColumns, err := tryGetStoredColumns(params.Storage, root, sortedSliceFromMap(storedIndices))
+		// Enough columns are available for reconstruction. Retrieve all stored columns.
+		allStoredColumns, err := storage.Get(root, nil)
 		if err != nil {
-			return nil, errors.Wrapf(err, "try get direct columns for root %#x", root)
+			return nil, errors.Wrapf(err, "failed to get all stored columns for reconstruction for block root %#x", root)
 		}
 
-		result[root] = append(result[root], storedColumns...)
+		// Attempt reconstruction.
+		reconstructedColumns, err := peerdas.ReconstructDataColumnSidecars(allStoredColumns)
```

Comment:
in that function why is the return called outVerifiedRoSidecars when we call it reconstructedColumns here